modules/core/src/main/java/org/apache/ignite/internal/QueryMXBeanImpl.java

```diff
@@ -0,0 +1,168 @@
+        A.notNull(routineId, "routineId");
+
+        cancelContinuous(UUID.fromString(routineId));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancelSQL(String id) {
+        A.notNull(id, "id");
+
+        if (log.isInfoEnabled())
+            log.info("Killing sql query[id=" + id + ']');
+
+        boolean res;
+
+        try {
+            IgniteClusterImpl cluster = ctx.cluster().get();
+
+            T2<UUID, Long> ids = parseGlobalQueryId(id);
+
+            if (ids == null)
+                throw new IllegalArgumentException("Expected global query id. " + EXPECTED_GLOBAL_QRY_ID_FORMAT);
+
+            res = cluster.compute().execute(new VisorQueryCancelTask(),
+                new VisorTaskArgument<>(ids.get1(), new VisorQueryCancelTaskArg(ids.get1(), ids.get2()), false));
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        if (!res)
+            throw new RuntimeException("Query not found[id=" + id + ']');
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancelScan(String originNodeId, String cacheName, Long id) {
+        A.notNullOrEmpty(originNodeId, "originNodeId");
+        A.notNullOrEmpty(cacheName, "cacheName");
+        A.notNull(id, "id");
+
+        if (log.isInfoEnabled())
+            log.info("Killing scan query[id=" + id + ",originNodeId=" + originNodeId + ']');
+
+        cancelScan(UUID.fromString(originNodeId), cacheName, id);
+    }
+
+    /**
+     * Kills scan query by the identifiers.
+     *
+     * @param originNodeId Originating node id.
+     * @param cacheName Cache name.
+     * @param id Scan query id.
+     */
+    public void cancelScan(UUID originNodeId, String cacheName, long id) {
+        boolean res;
+
+        try {
+            IgniteClusterImpl cluster = ctx.cluster().get();
+
+            ClusterNode srv = U.randomServerNode(ctx);
+
+            IgniteCompute compute = cluster.compute();
+
+            res = compute.execute(new VisorScanQueryCancelTask(),
+                new VisorTaskArgument<>(srv.id(),
+                    new VisorScanQueryCancelTaskArg(originNodeId, cacheName, id), false));
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        if (!res) {
+            throw new RuntimeException("Query not found[originNodeId=" + originNodeId +
+                ",cacheName=" + cacheName + ",qryId=" + id + ']');
+        }
+    }
+
+    /**
+     * Kills continuous query by the identifier.
+     *
+     * @param routineId Routine id.
+     */
+    public void cancelContinuous(UUID routineId) {
+        if (log.isInfoEnabled())
+            log.info("Killing continuous query[routineId=" + routineId + ']');
+
+        boolean res;
+
+        try {
+            IgniteClusterImpl cluster = ctx.cluster().get();
+
+            IgniteCompute compute = cluster.compute();
+
+            ClusterNode srv = U.randomServerNode(ctx);
```

Comment:
Why random node? Why not send it to the initiator?

Reply:
> Why it's not a good idea to send cancel message to the initiator

My opinion - canceling of the query is the "last line of defense".
When an administrator decides to cancel a query it means something goes wrong.
The whole initiator node can be flooded with the entries update.
So I think it's not the best choice to send a cancel request to the initiator.

> Why random node?

CQ cancel implemented via discovery message so there is no difference what specific node receives request first.

Reply:
I think the cancellation task should be delegated to the initiator. See the explanation below.